### DISCUSSION TITLE
PURL changes after SBOM load?

### DISCUSSION THREAD

**Discussion by kbolander** [DependencyTrack/dependency-track#1456] [Q&A]
This might be an issue, but wanted to ask the question first in case I'm missing something. I have an SBOM where a given package has a package url of `pkg:maven/org.keycloak.testsuite/integration-arquillian-tests-base@10.0.2?type=jar` however, when it's actually stored in the system, it's getting stored as `pkg:maven/org.keycloak.testsuite/integration-arquillian-tests-base@10.0.2?classifier=tests&type=test-jar`. You'll notice the extra info in the qualifiers/subpath sections. 

I can see in the [package url code](https://github.com/package-url/packageurl-java) that there appears to be some modifications that the purl strings are run through, but I'm trying to figure out why this would be modifying the purl in the capacity I'm seeing? Is this expected behavior or is this maybe a bug? Thank you!

**EDIT**
I will add that, it only appears to be getting changed upon reloading a BOM to an existing project. For example, the first load of the BOM to a net new project shows the purl as I'd expect -- matching the purl as it appears in the BOM. When, I go to reupload the BOM to the existing project, that's when I see the purl change and no longer match. This is starting to feel more like a bug, but still curious if anyone has thoughts/ideas as to what could be happening?

**Comment by stevespringett** [comment]
I uploaded the following multiple times, but the purl does not change. It remains the same, as it should.

```json
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.2",
  "serialNumber": "urn:uuid:371ffb8c-c11e-42b5-b5b9-9280fc62783e",
  "version": 1,
  "components": [
    {
      "type": "library",
      "bom-ref": "pkg:maven/org.keycloak.testsuite/integration-arquillian-tests-base@10.0.2?type=jar",
      "group": "org.keycloak.testsuite",
      "name": "integration-arquillian-tests-base",
      "version": "10.0.2",
      "purl": "pkg:maven/org.keycloak.testsuite/integration-arquillian-tests-base@10.0.2?type=jar"
    }
  ]
}
```

I'm wondering if some pom may have a test scope and in other poms it does not.  If you're running an aggregate goal, I guess its possible that one module may have this component as test scope and another module does not. That's the only thing I can think of.

**Reply by kbolander** [reply]
I probably should have added this before, but I am using the API call to upload the SBOM -- not sure if that changes things. I'm still able to consistently reproduce using the `api/v1/bom` API call. I'm passing the file in like so: `response = requests.post(f"{baseurl}bom", headers=headers, files=files)` where:

 `headers = {"X-Api-Key": dt_api_key}`

```
    files = {
        "autoCreate": (None, "true"),
        "projectName": (None, proj_name),
        "projectVersion": (None, proj_version),
        "bom": (file_path, open(file_path, "rb")),
    }
```

I'm using this same JSON file each time (note the attached is a .txt version since I can't upload it as a .json)
[Valid_SBOM_CDX_2.txt](https://github.com/DependencyTrack/dependency-track/files/8215375/Valid_SBOM_CDX_2.txt)

Not sure if that changes anything, but thought I'd add the additional detail for context and hx reference.

**Reply by kbolander** [reply]
And now I looked at this file again and it looks like there are actually 2 packages marked on this, one has purl: `pkg:maven/org.keycloak.testsuite/integration-arquillian-tests-base@10.0.2?classifier=tests&type=test-jar` and the other has purl: `pkg:maven/org.keycloak.testsuite/integration-arquillian-tests-base@10.0.2?type=jar`

But, what's weird is, when the BOM is first loaded, both packages are loaded like so. Notice total components are 903:
![image](https://user-images.githubusercontent.com/96793624/157475135-f030d4be-27bd-46b0-8558-74e94fcbd90f.png)
And you can see both line items for the integration-arquillian-tests-base:
![image](https://user-images.githubusercontent.com/96793624/157475251-631af4ac-c041-438a-b993-17a339761347.png)

However, after I reupload the same BOM file to the project, one of these components is dropped. Notice total components are 902:
![image](https://user-images.githubusercontent.com/96793624/157475355-f96f34f8-a4d3-4fb8-a06e-34ab739b6100.png)
And now there is only one line item for the integration-arquillian-tests-base:
![image](https://user-images.githubusercontent.com/96793624/157475458-cb41a818-bb01-4760-9f46-9cfed95f1056.png)
